components/nodemanager-service/pgdb/migration/sql/07_upsert_functions.up.sql

```diff
@@ -0,0 +1,60 @@
+CREATE OR REPLACE FUNCTION upsert_by_source_id_run_data(_id text, _name text, _platform text, _platform_version text, _source_state text,
+		_last_contact timestamp, _source_id text, _source_region text, _source_account_id text, _last_run json, _projects_data json, _manager text)
+RETURNS TEXT
+LANGUAGE plpgsql
+AS $$
+BEGIN
+    UPDATE nodes SET name = name, platform = _platform, platform_version = _platform_version, source_state = _source_state,
+        last_contact = _last_contact,  source_id = _source_id, source_region = _source_region, source_account_id = _source_account_id, last_run = _last_run, projects_data = _projects_data
+        WHERE id = _id;
+    IF found THEN
+        RETURN _id;
+    END IF;
```

Comment:
First, we try to update with the ID. If there is a row matching that ID we return.